channeldb/control_tower.go

```diff
@@ -0,0 +1,394 @@
+		}
+		attemptBytes = a.Bytes()
+	}
+
+	switch paymentStatus {
+
+	// We already have an InFlight payment on the network. We will disallow
+	// any first payment attempts, but allow updating new attempts.
+	case StatusInFlight:
+		if firstAttempt {
+			return ErrPaymentInFlight
+		}
+
+	// We allow retrying failed payments.
+	case StatusFailed:
+
+	// It is safe to reattempt a payment if we know that we haven't left
+	// one in flight.
+	case StatusGrounded:
+		// If the payment is grounded, this must be the first
+		// attempt, or else it would have a status already in
+		// the DB.
+		if !firstAttempt {
+			return ErrPaymentNotInitiated
+		}
+
+	// We've already completed a payment to this payment hash,
+	// forbid the switch from sending another.
+	case StatusCompleted:
+		return ErrAlreadyPaid
+
+	default:
+		return ErrUnknownPaymentStatus
+	}
+
+	// We'll move it into the InFlight state.
+	paymentStatus = StatusInFlight
+	err := bucket.Put(paymentStatusKey, StatusInFlight.Bytes())
+	if err != nil {
+		return err
+	}
+
+	// Now record the information in the DB if this is the first attempt.
+	if firstAttempt {
+		// Add the payment info to the bucket, which contains the
+		// static information for this payment
+		err := bucket.Put(paymentCreationInfoKey, infoBytes)
+		if err != nil {
+			return err
+		}
+
+		// We'll add an empty attempt info to start with, in case we
+		// are initializing a payment that went straight to
+		// Failed/Success without an attempt being crafted.
+		err = bucket.Put(paymentAttemptInfoKey, attemptBytes)
+		if err != nil {
+			return err
+		}
+
+		// Since this payment is not yet settled, we'll record an all-
+		// zero preimage.
+		var zero lntypes.Preimage
+		err = bucket.Put(paymentSettleInfoKey, zero[:])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RegisterAttempt atomically moves the payment into the InFlight state. The
+// provided AttemptInfo is recorded in the DB. This method checks that no
+// completed payment exist for this payment hash.
+func (p *paymentControl) RegisterAttempt(paymentHash lntypes.Hash,
+	info *CreationInfo, attempt *AttemptInfo) error {
+
+	// Serialize the information before opening the db transaction.
+	var a bytes.Buffer
+	if err := serializeAttemptInfo(&a, attempt); err != nil {
+		return err
+	}
+	attemptBytes := a.Bytes()
+
+	return p.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		// Init the payment to make sure we are not owerwriting a n
+		// InFlight attempt if we expect this to be the first attempt.
+		err = initPayment(bucket, paymentHash, info)
+		if err != nil {
+			return err
+		}
+
+		// We already have an InFlight payment on the network. We will
+		// disallow any first payment attempts, but allow updating the
+		// until a response is received.
+		// Add the payment attempt to the payments bucket.
+		err = bucket.Put(paymentAttemptInfoKey, attemptBytes)
```

Comment:
Is it written to the db twice for first attempts? (first time inside `initPayment`)